src/ppx/grfx/grfx_swapchain.cpp

```diff
@@ -402,5 +573,93 @@ Result Swapchain::PresentHeadless(uint32_t imageIndex, uint32_t waitSemaphoreCou
     return ppx::SUCCESS;
 }
 
+void Swapchain::RecordUI(uint32_t imageIndex, std::function<void(grfx::CommandBufferPtr)> f)
```

Comment:
A bit confused here, why is RecordUI part of Swapchain?

Reply:
Currently ImGui is rendered in the middle of overall rendering process, which means any timing of command buffer will include the ImGui, and if we set scissor or render to a larger than screen buffer it will be truncated or have incorrect size.
Logically it should not be a part of Swapchain, and be part of application post render processing step, however it happens between app finish rendering (presenting) and send the image to display, which currently all resides in Swapchain. It's a way larger refactoring to make render not operate directly on a swapchain :(

Reply:
i couldn't find where is this function called, but i guess it is supposed to be called in the Swapchain::Present? I am wondering would it be cleaner if we keep a list of callback functions as member of the swapchain, and let the application to register callbacks (for example, UI callback), and we check the size of the registered callbacks, and only call this function (probably rename it to something else, something like `ExecuteCallbacksAfterBlit`) when the size of the callbacks are not 0?

Reply:
It's not... yet, I'm rebasing a [giant PR](https://github.com/hysw/bigwheels/tree/tianc-vswap-offscreen) I had previously.
Set a callback by application is a better idea, except I need to re-architect how UI is drawn (which is currently in the middle of frame) and also make sure it does not break XR, which will take a bit more changes.

Reply:
I am ok with submitting this temp solution if the real solution is coming soon. Also please add some comments to explain this with a link to the bug in the buganizer so that we don't forget about it. :)